### PR TITLE
Modem backend uart async optimization

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
The mode UART backend implementations have been manually tested ad-hoc up until now. This PR adds a proper test suite to stress test the backends on real hardware, along with patches to the backends fixing issues found by the new tests :)

The patches to the backends are listed here, and have their own commits in this PR:
1. When pipe close is invoked, the async UART backend will now wait for all buffers released, TX aborted, and RX disabled, before indicating it has been closed. This ensures the UART device is in a known state once the pipe is closed. In addition, the close invokation will also call uart_tx_abort().
2. Use single RX ring buffer protected by spinlock instead of switching between two RX smaller ring buffers. The tests revealed there was no measurable performance over using a single buffer larger RX buffer, which is simpler.
3. Flushing the UART is not the responsibility of the async UART backend when using the async UART API, so it has been removed.

### CODE REVIEW

File: subsys/modem/backends/modem_backend_uart_async.c [hunk 1]
```diff
@@ -153,12 +144,15 @@ static int modem_backend_uart_async_open(void *data)
 	 * used to store received data.
 	 */
 	ret = uart_rx_enable(backend->uart, backend->async.receive_bufs[0],
-			     backend->async.receive_buf_size, 3000);
+			     backend->async.receive_buf_size, 30000);
```
Comment (rerickson1) [inline subsys/modem/backends/modem_backend_uart_async.c:147]: Should there be a define or kconfig for this value?

File: tests/subsys/modem/backends/uart/boards/nrf5340dk_nrf5340_cpuapp.overlay [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+&pinctrl {
+	uart1_default: uart1_default {
+		group1 {
+			psels = <NRF_PSEL(UART_TX, 1, 11)>;
+		};
+		group2 {
+			psels = <NRF_PSEL(UART_RX, 1, 10)>;
+			bias-pull-up;
```
Comment (rerickson1) [inline tests/subsys/modem/backends/uart/boards/nrf5340dk_nrf5340_cpuapp.overlay:27]: Technically you shouldn't need a pull up here...

File: tests/subsys/modem/backends/uart/src/main.c [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+		.uart = uart,
+		.receive_buf = backend_receive_buffer,
+		.receive_buf_size = 1024,
+		.transmit_buf = backend_transmit_buffer,
+		.transmit_buf_size = 1024,
+	};
+
+	pipe = modem_backend_uart_init(&uart_backend, &config);
+	modem_pipe_attach(pipe, modem_pipe_callback_handler, NULL);
+	k_msleep(10);
```
Comment (rerickson1) [inline tests/subsys/modem/backends/uart/src/main.c:167]: I never like time delays in code. Do we have to have this? If so, add a comment why.

**Review by rerickson1 — CHANGES_REQUESTED** [review]

File: tests/subsys/modem/backends/uart/src/main.c (hunk 3 above)
Comment (bjarki-trackunit) [reply tests/subsys/modem/backends/uart/src/main.c]: This was a leftover from me trying to solve this bug #65320, I have removed the timeout :)

File: subsys/modem/backends/modem_backend_uart_async.c (hunk 1 above)
Comment (bjarki-trackunit) [reply subsys/modem/backends/modem_backend_uart_async.c]: Added a Kconfig for this value

File: tests/subsys/modem/backends/uart/boards/nrf5340dk_nrf5340_cpuapp.overlay (hunk 2 above)
Comment (bjarki-trackunit) [reply tests/subsys/modem/backends/uart/boards/nrf5340dk_nrf5340_cpuapp.overlay]: Tested and verified that it is not needed, and have removed it :)

**Review by rerickson1 — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**